sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
 
-    outfile = 'spdfa-clustered-'+datafile
-    os.system("dot -Tpng "+outfile+"-dfa.dot -o "+outfile+"-dfa.png")
+    os.system("dot -Tpng " + filename + ".dot -o " + filename + ".png")
     return state_groups
 
-def make_av_data(condensed_data):
-## Experiment: attack graph for one victim w.r.t time
-    condensed_v_data= dict()
-    for attacker,episodes in condensed_data.items():
-        team = attacker.split('-')[0]
-        victim_ = attacker.split('->')[1]
-        tv = team+'-'+victim_
-        #print(tv)
-        if tv not in condensed_v_data.keys():
-            condensed_v_data[tv] = []
-        condensed_v_data[tv].extend(episodes)
-        condensed_v_data[tv] = sorted(condensed_v_data[tv], key=lambda item: item[0])
-    condensed_v_data = {k: v for k, v in sorted(condensed_v_data.items(), key=lambda item: len([x[0] for x in item[1]]))}
-    #print([(k,len([x[0] for x in v])) for k,v in condensed_v_data.items()])
-    print('Victims hosts: ', (set([x.split('-')[-1] for x in condensed_v_data.keys()])))
-
-    condensed_a_data= dict()
-    for attacker,episodes in condensed_data.items():
-        team = attacker.split('-')[0]
-        attacker_ = (attacker.split('->')[0]).split('-')[1]
-        tv = team+'-'+attacker_
-        #print(tv)
-        if tv not in condensed_a_data.keys():
-            condensed_a_data[tv] = []
-            
-        condensed_a_data[tv].extend(episodes)
-        condensed_a_data[tv] = sorted(condensed_a_data[tv], key=lambda item: item[0])
-        #print(len(condensed_a_data[tv]))
-    #condensed_a_data = {k: v for k, v in sorted(condensed_a_data.items(), key=lambda item: item[1][0][0])}
-    #print([(k,[x[0] for x in v]) for k,v in condensed_a_data.items()])
-    print('Attacker hosts: ', (set([x.split('-')[1] for x in condensed_a_data.keys()])))
-    return (condensed_a_data, condensed_v_data)
-
-## translate technical nodes to human readable
-def translate(label, root=False):
+
+# Step 6.4: Group episodes (state sequences) per (team, victim) pair
+def group_episodes_per_av(state_sequences):
+    # Experiment: attack graph for one victim w.r.t time
+    victim_episodes = dict()  # Episodes per (team, victim)
+    for attack, episodes in state_sequences.items():
+        team = attack.split('-')[0]
+        victim = attack.split('->')[1]
+        team_victim = team + '-' + victim
+        if team_victim not in victim_episodes.keys():
+            victim_episodes[team_victim] = []
+        victim_episodes[team_victim].extend(episodes)
+        victim_episodes[team_victim] = sorted(victim_episodes[team_victim], key=lambda epi: epi[0])  # By start time
+    # Sort by start time across all
+    victim_episodes = {k: v for k, v in sorted(victim_episodes.items(), key=lambda kv: len([epi[0] for epi in kv[1]]))}
+    print('Victims hosts: ', set([team_victim.split('-')[-1] for team_victim in victim_episodes.keys()]))
+
+    attacker_episodes = dict()  # Episodes per (team, attacker)
```

Comment:
Attacker data is not used anywhere. Still keeping it?

Reply:
Not used but keep it still